### PR TITLE
Fixed issue #2227

### PR DESCRIPTION
**Pull request by vtellier**
# Problem
See #2227 
ZMQ doesn't provide status about the encryption handshake. When it fails, the client tries to reconnect indefinitely.

# Solution
Added two new monitoring events:
 - **ZMQ_EVENT_HANDSHAKE_SUCCEED** is raised once the encryption handshake succeed
 - **ZMQ_EVENT_HANDSHAKE_FAILED** is raised when it failed
Both events are raised on server and client side.

Also, when the curve engine find out the client uses a different (server) public key, it sends back a hard error to the client so it doesn't tries to reconnect infinitely.
That error message contains an error code abitrarily set to 100:

```
// src/curve_server.cpp
status_code = "100";
```

This can (or should) be modified, as the error code is not read so far, I had to set a value.

_**Nb:** I compiled and tested my code under **Visual Studio 2015** (non XP), I did **not** tested with any language binding or other zmq projects._

### CODE REVIEW

File: include/zmq.h [hunk 1]
```diff
@@ -417,6 +417,8 @@ ZMQ_EXPORT const char *zmq_msg_gets (zmq_msg_t *msg, const char *property);
 #define ZMQ_EVENT_CLOSE_FAILED      0x0100
 #define ZMQ_EVENT_DISCONNECTED      0x0200
 #define ZMQ_EVENT_MONITOR_STOPPED   0x0400
+#define ZMQ_EVENT_HANDSHAKE_FAILED  0x0800
```
Comment (bluca) [inline include/zmq.h]: Please move these in the DRAFT section at the bottom of the file (and copy them in src/zmq_draft.hpp too)

File: src/curve_server.cpp [hunk 2]
```diff
@@ -310,13 +310,14 @@ int zmq::curve_server_t::process_hello (msg_t *msg_)
                               sizeof hello_box,
                               hello_nonce, cn_client, secret_key);
     if (rc != 0) {
-        //  Temporary support for security debugging
-        puts ("CURVE I: cannot open client HELLO -- wrong server key?");
-        errno = EPROTO;
-        return -1;
+        // Hard error, the client knows a wrong server public key, it shall not try to reconnect using the same.
```
Comment (bluca) [inline src/curve_server.cpp]: Please revert this change - as mentioned in https://github.com/zeromq/libzmq/issues/2227#issuecomment-269786977 this is not backward compatible

File: src/socket_base.cpp [hunk 3]
```diff
@@ -1678,6 +1678,16 @@ void zmq::socket_base_t::event_disconnected (const std::string &addr_, zmq::fd_t
     event(addr_, fd_, ZMQ_EVENT_DISCONNECTED);
 }
 
+void zmq::socket_base_t::event_handshake_failed(const std::string &addr_, int err_)
+{
+	event(addr_, err_, ZMQ_EVENT_HANDSHAKE_FAILED);
```
Comment (bluca) [inline src/socket_base.cpp]: The formatting in the library is spaces-only, 4 spaces indentation, please fix this and other tabs in this diff

**Comment by bluca** [comment]
The test_monitor unit test is broken by this change, could you please have a look and fix it?

See https://travis-ci.org/zeromq/libzmq/builds/187723849

**Comment by bluca** [comment]
@vtellier ping - are you looking into this?

**Comment by vtellier** [comment]
Yes! Sorry I was away, back online in few minutes and fix it

**Comment by bluca** [comment]
Thank you!
